### PR TITLE
feat(conditions): add current-day/month/year/day-of-week conditions

### PR DESCRIPTION
**Pull request by sileht**
Change-Id: Id6407aaff6e4ad76f55f7a12770a07dbd7cf6769

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
1. feat(rules): allow and/or in conditions ([#2691](https://github.com/Mergifyio/mergify-engine/pull/2691))
1. feat: allow merged/closed/author/base/head in and/or conditions ([#2692](https://github.com/Mergifyio/mergify-engine/pull/2692))
1. feat(conditions): allow checks condition in and/or ([#2693](https://github.com/Mergifyio/mergify-engine/pull/2693))
1. refactor(filter): make it generic ([#2694](https://github.com/Mergifyio/mergify-engine/pull/2694))
1. feat(time): add time conditions ([#2695](https://github.com/Mergifyio/mergify-engine/pull/2695))
1. feat(conditions): add day/month/yeat/day-of-week conditions ([#2705](https://github.com/Mergifyio/mergify-engine/pull/2705))

File: mergify_engine/rules/parser.py [hunk 1]
```diff
@@ -45,6 +47,39 @@
...
+        lambda tokens: tokens[0].value >= 1 and tokens[0].value <= 12,
+        message="month must be between 1 and 12",
+    )
+)
+_year = (
+    pyparsing.Word(pyparsing.nums)
+    .setParseAction(lambda tokens: date.Year(int(tokens[0])))
+    .addCondition(
+        lambda tokens: tokens[0].value >= 1900 and tokens[0].value <= 9999,
+        message="year must be between 1900 and 9999",
```
Comment (jd) [inline mergify_engine/rules/parser.py:71]: not sure anything before epoch would work
Go ahead with 2000 🤷

File: mergify_engine/rules/parser.py [hunk 2]
```diff
@@ -45,6 +47,39 @@
...
+    )
+)
+_day_of_week = (
+    pyparsing.Word(pyparsing.nums)
+    .setParseAction(lambda tokens: date.DayOfWeek(int(tokens[0])))
+    .addCondition(
+        lambda tokens: tokens[0].value >= 1 and tokens[0].value <= 7,
+        message="day-of-week must be between 1 and 7",
```
Comment (jd) [inline mergify_engine/rules/parser.py:79]: Can you check these messages are printed correctly?

File: mergify_engine/tests/unit/rules/test_filter.py [hunk 3]
```diff
@@ -215,6 +217,191 @@ async def test_time_binary() -> None:
...
+
+    f = filter.BinaryFilter({">": ("foo", klass(5))})
+    assert await f(FakePR({"foo": klass(30)}))
+    assert not await f(FakePR({"foo": klass(2)}))
+    assert not await f(FakePR({"foo": klass(5)}))
+
+
+@freeze_time("2012-01-14T12:15:00")
+async def test_day_near_datetime() -> None:
+    nextday = datetime.datetime(2012, 1, 7, 0, 0, 0, 0, tzinfo=datetime.timezone.utc)
```
Comment (jd) [inline mergify_engine/tests/unit/rules/test_filter.py:254]: I'm a bit confused: this says next day is the 7th January but the time is frozen on 14th January? 🤔

**Review by jd — CHANGES_REQUESTED** [review]

File: docs/source/conditions.rst [hunk 4]
```diff
@@ -255,6 +255,18 @@ Here's the list of pull request attribute that can be used in conditions:
...
+     - The current day of the month, from 1 to 31.
+   * - ``current-month``
+     - integer
+     - The current month, from 1 to 12.
+   * - ``current-year``
+     - integer
+     - The current year, from 1900 to 9999.
+   * - ``current-day-of-week``
+     - integer
+     - The current day of the week, from 1 (Monday) to 7 (Sunday).
```
[the remark this answers is not in the document]
Comment (sileht) [reply docs/source/conditions.rst:269]: String support is here: https://github.com/Mergifyio/mergify-engine/pull/2708

File: mergify_engine/rules/filter.py [hunk 5]
```diff
@@ -343,29 +363,65 @@ def _dt_op(
...
+                        dt_ref = dt_ref + datetime.timedelta(days=1)
+                    elif isinstance(ref, date.Day):
+                        dt_ref = dt_ref + datetime.timedelta(days=1)
+                    elif isinstance(ref, date.Month):
+                        dt_ref = dt_ref.replace(day=1)
+                        dt_ref = dt_ref + datetime.timedelta(days=32)
+                        dt_ref = dt_ref.replace(day=1)
+                    elif isinstance(ref, date.Year):
+                        dt_ref = dt_ref.replace(month=1, day=1)
+                        dt_ref = dt_ref + datetime.timedelta(days=370)
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/rules/filter.py:389]: We want the next 1st January, so I add a bit more than one Year (~365 days) and pick 370 and replace the month and the day.

Then I just read the iso8601 standard and found: `the 'ISO day of year' numbers, this is a number between 1 and 371` :facepalm:

[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/rules/filter.py:389]: anyway 366 is enough

File: mergify_engine/tests/unit/rules/test_filter.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/tests/unit/rules/test_filter.py:254]: The test just need the year and month to be frozen, to check override the day. I put 6 by default anyway.

File: mergify_engine/rules/parser.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/rules/parser.py:79]: This message can show up only in tests. When we parse the configuration you never get the level of details. Pyparsing just prints the expression is expected. If this one fails, it tries the next one and returns a very long generic error if none matches. This error contains all possibilities.

Comment (jd) [reply mergify_engine/rules/parser.py:79]: That's what I thought unfortunately 😢

File: docs/source/conditions.rst [hunk 6]
```diff
@@ -255,6 +255,19 @@ Here's the list of pull request attribute that can be used in conditions:
...
+   * - ``current-month``
+     - integer
+     - The current month, from 1 to 12.
+   * - ``current-year``
+     - integer
+     - The current year, from 1900 to 9999.
+   * - ``current-day-of-week``
+     - integer
+     - The current day of the week, week start on Monday, from 1 (Monday) to 7
+       (Sunday).
```
Comment (jd) [inline docs/source/conditions.rst:270]:
```suggestion
     - The current day of the week. From 1 (Monday) to 7
       (Sunday). Weeks start on Monday.
```

**Review by jd — CHANGES_REQUESTED** [review]

**Review by jd — APPROVED** [review]

**Pull request merged**